### DISCUSSION TITLE
Detect holes in the myelin

### DISCUSSION THREAD

**Discussion by hermancollin** [neuropoly/axondeepseg#812] [Ideas of new features]
Continuation of a discussion started by @lgazdzin in #776 about detecting how loose/compact myelin sheaths. Here are @lgazdzin's thoughts:

> We think it's useful to not fill in the 'holes' in the myelin because then the myelin thickness would be sort of artificially thick in some cases. We think it would be more valid to get a measure of the myelin thickness that just includes actual myelin rather than the total space taken up by myelin around a given axon. But then having an idea of how loose the myelin is would be of potential interest.
> My first attempt at this was to calculate an 'expected' perimeter for the axonmyelin object using the minor axis (i.e. axon diameter + 2X myelin thickness) and the axonmyelin area (to calculate the major axis of the ellipse). I then take the ratio of this ADS reported axonmyelin perimeter to this expected perimeter to get a sense for how much extra perimeter there is due to holes in the axonmyelin segmentation. This approach seems promising in that it seems to pick up on the fact that the optic tracts in our images have 'looser' myelin than the corpus callosum and anterior commissure, but it's a bit confusing because the value of the metric is actually <1 in most cases.
>Another issue with this is that it is oblivious to the fact that the area contains the axon, so the metric is not specific to the myelin object. It might be better if we could get the perimeter of the myelin object separately. I realize that this would then include the axon perimeter as well as the inner edge of the myelin, but that might actually be a good thing because it would allow the metric to 'know about' the axon.
> Another idea I had was to maybe use the eccentricity and solidity measures to somehow 'weight' or normalize the ADS perimeter to try to isolate the 'extra' perimeter to the holes in the myelin. My idea was to use the ratio of the (weighted; normalized) perimeter to the area as a measure of myelin looseness. My theory here is that the perimeter:area ratio could be greater than 'expected' due to 1) smaller solidity (more undulations or bumps around the edge), 2) greater eccentricity (more elongated ellipse would have greater perimeter:area than a circle), or more holes in the segmentation. If we could somehow correct for the contribution of solidity and eccentricity, we could maybe isolate to the effect of the holes.

Expanding on your last idea, I think the simplest way to quantify myelin looseness would be to measure the area of the holes in the myelin itself. What do you think? Instead of estimating this area, we could actually compute it.

Here is how it would work:
1. For one axon, select the axon and myelin pixels, like so: 
![select_pixels](https://github.com/axondeepseg/axondeepseg/assets/83031821/bb667701-73b8-42bb-91de-67305589c937)
2. Then, we apply [skimage.morphology.remove_small_holes](https://scikit-image.org/docs/stable/api/skimage.morphology.html#skimage.morphology.remove_small_holes) on this region. This function accepts an area threshold, under which all holes will be filled. We can set this threshold to the axon area (realistically the myelin should never have holes bigger than the axon itself). This will give us the following: 
![remove_holes](https://github.com/axondeepseg/axondeepseg/assets/83031821/c8f3316a-88bd-4899-83d7-6dbaf9ea32ce)
3.  We compute this `full_area`. Then, `holes_area` = `full_area` - `axonmyelin_area`. 

@lgazdzin would the "hole area" inside the myelin allow you to perform your analysis?

**Comment by lgazdzin** [comment]
@hermancollin I think this is a great idea. One consideration is that the axon area also sometimes has holes in the segmentation. It would be ideal to fix these holes first (which would also potentially change the axon diameter estimation, so if this isn't already in the pipeline, it would potentially be a good addition) so that they don't contribute to the myelin looseness metric. .

**Reply by hermancollin** [reply]
Good point. The axon class should never contain holes in principle, so this has to be accounted for. We can do this easily by applying the remove_small_holes function directly on the axon mask.

@lgazdzin what do you think of holes between the axon and the myelin? In the mask I included above, we can actually see this phenomenon on the bottom-left. We know from experience that this happens sometimes. Should we include or exclude these regions? I don't have an easy fix for this off the top of my head so I'm not sure how much it would complicate the pipeline.
![Screenshot_20240705_133141](https://github.com/axondeepseg/axondeepseg/assets/83031821/35f69d4b-9261-4687-87ff-80a33c401478)

**Reply by lgazdzin** [reply]
I'm assuming these spaces are usually pretty small. My inclination would be to ignore them for now. I guess there are maybe two ways of dealing with this: 1) fill in the holes in the axon object alone and then do as you suggest above to measure the remaining holes in the axon-myelin object, which would include these spaces or 2) instead of measuring the holes in the axon-myelin object, can the holes in the myelin object be measured instead (which would then not include these spaces)?

**Reply by lgazdzin** [reply]
Also, in some cases, these spaces may actually be 'real'. i.e. they may be the 'periaxonal space', which is between the axon and the myelin.

**Comment by hermancollin** [comment]
@lgazdzin I don't have the bandwidth to properly integrate this into ADS at the moment, but I experimented with this feature in #814. You can checkout the `ac/myelin_holes_area` working branch and see if this is something you could use. Using a new `-d` option, the morphometrics script now saves a new column with the area of the holes in the myelin. If you have more specific ideas to improve this feature, please let me know.

**Comment by lgazdzin** [comment]
Thanks Armand! I'll take a look at this and let you know how it goes!
________________________________
